### PR TITLE
[WIP] DigDog Features for weighted constants and sequences.

### PR DESCRIPTION
**Pull request by jkotalik**
Contains three main features: 
Fix to #150 
Implements weighted constants using a modified tf-idf formula.
Implements weighted sequence selection based on runtime, length, and number of times run.

We have written a report on our findings on these features. From our results, these features look promising.

[digdog-improving-randoops.pdf](https://github.com/randoop/randoop/files/837095/digdog-improving-randoops.pdf)

We need to discuss if:
1. Would we like these features in Randoop?
2. How would we like to merge these features with Randoop's already existing features.
3. Any design changes that should be made because of these features.

### CODE REVIEW

File: src/main/java/randoop/JunitFileWriter.java [hunk 1]
```diff
@@ -351,46 +352,61 @@ public File writeDriverFile() {
...
+            out.println();
+            out.println("  public static void " + classMethodName + "() {");
+            totalCallCount = 0;
           }
         }
 
         if (afterAllText != null) {
           out.println("    " + testClass + "." + AFTER_ALL_METHOD + "();");
         }
-
```
Comment (mernst) [inline src/main/java/randoop/JunitFileWriter.java]: Could you please remove unnecessary changes from your fork?
That will make the diffs smaller to read and, once merged, will make the version control history smaller and cleaner.
Thanks!

File: src/main/java/randoop/generation/ForwardGenerator.java [hunk 2]
```diff
@@ -482,7 +482,7 @@ private void randoopConsistencyTest2(Sequence newSequence) {
 
   // Checks that the set allSequencesAsCode contains a set of strings
   // equivalent to the sequences in allSequences.
-  private void randoopConsistencyTests(Sequence newSequence) {
+  protected void randoopConsistencyTests(Sequence newSequence) {
```
Comment (mernst) [inline src/main/java/randoop/generation/ForwardGenerator.java]: Why has this been changed from private to protected?  It doesn't seem to be called anywhere except from this file.
Can you either revert all these changes (which will make the diffs smaller), or create a separate pull request so that each pull request has a single purpose and can have a descriptive commit message?

File: src/main/java/randoop/generation/WeightedComponentManager.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package randoop.generation;
+
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import randoop.main.GenInputsAbstract;
+import randoop.sequence.Sequence;
+
+/**
+ * WeightedComponentManager extends the functionality of ComponentManager by adding needed
```
Comment (mernst) [inline src/main/java/randoop/generation/WeightedComponentManager.java]: "needed" is not very descriptive here.  Could you document what this class does?  I think you can do so in just a sentence or two, and that will make this code stand on its own.

File: src/docs/manual/index.html [hunk 4]
```diff
@@ -632,119 +632,90 @@ <h3 id="command-line-options">Command-line options</h3>
...
  <pre class="code">
  cons : Type0.&lt;init&gt;(Type1, Type2, ..., TypeN)
  method : Type0.method_name(Type1, Type2, ..., TypeN)
  </pre>
 
- <p>
- Each <code>Type<i>i</i></code> must be fully-qualified (include package
- names).
- </p>
+ <p>Each <code>Type<i>i</i></code> must be fully-qualified (include package names).
```
Comment (mernst) [inline src/docs/manual/index.html]: There are many gratuitous changes to the manual (the index.html file), such as this one.  Please undo all of them.  They make it hard to find the real changes in the diffs, and they will pollute the version control history if merged into master.

File: src/main/java/randoop/generation/WeightedComponentManager.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * functionality for weighted sequences and constants, for the command-line options
+ * --weighted-sequences and --weighted-constants.
+ */
+public class WeightedComponentManager extends ComponentManager {
+
+  /** Sequence frequency represents the number of times a sequence occurs in a set of classes */
```
Comment (mernst) [inline src/main/java/randoop/generation/WeightedComponentManager.java]: There seems to be a type error here.  A sequence cannot occur in a set of classes -- only a class can appear in a set of classes.  Can you clarify the comment?

File: src/main/java/randoop/main/GenInputsAbstract.java [hunk 6]
```diff
@@ -431,6 +447,14 @@ public GenInputsAbstract(
   public static boolean small_tests = false;
 
   /**
+   * Whether to use DigDog weighted sequences in sequence selection. Note that this weighting scheme
```
Comment (mernst) [inline src/main/java/randoop/main/GenInputsAbstract.java]: "DigDog weighted sequences" is not defined.  It needs an explanation, either here or in the manual.  The manual needs to be self-contained, preferably without reference to external documents.

File: src/main/java/randoop/main/GenTests.java [hunk 7]
```diff
@@ -453,10 +522,55 @@ public boolean handle(String[] args) throws RandoopTextuiException {
     if (!GenInputsAbstract.noprogressdisplay) {
       System.out.printf("%nInvalid tests generated: %d", explorer.invalidSequenceCount);
     }
+    // TODO: output to file for sequence comparison between DigDog/Randoop
+    if (GenInputsAbstract.output_sequence_info) {
+      DigDogGenerator fExplorer =
+          (DigDogGenerator) explorer; // should work, explorer should always be a forw.gen.
```
Comment (mernst) [inline src/main/java/randoop/main/GenTests.java]: Comment seems extraneous -- a programmer wouldn't write code that isn't intended to work.  I would cut the comment.

File: src/main/java/randoop/main/GenTests.java [hunk 8]
```diff
@@ -453,10 +522,55 @@ public boolean handle(String[] args) throws RandoopTextuiException {
...
+      Map<Sequence, List<Integer>> debugMap = fExplorer.getSequenceSizeMap();
+      writeTestInfo(debugMap);
+    }
+
     return true;
   }
 
   /**
+   * Write each sequence's info out to GenInputsAbstract.output_sequence_info_filename in .csv
```
Comment (mernst) [inline src/main/java/randoop/main/GenTests.java]: What is the info?  What is the format of the resulting .csv file?

File: src/main/java/randoop/reflection/AbstractOperationModel.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import randoop.generation.ComponentManager;
+import randoop.main.GenInputsAbstract;
+import randoop.operation.*;
+import randoop.sequence.Sequence;
+import randoop.test.ContractSet;
+import randoop.types.ClassOrInterfaceType;
+import randoop.types.Type;
+import randoop.util.MultiMap;
+
+/** Created by Michael on 3/3/2017. */
```
Comment (mernst) [inline src/main/java/randoop/reflection/AbstractOperationModel.java]: Please write documentation about the class and what it represents.

File: src/main/java/randoop/reflection/ClassLiteralExtractor.java [hunk 10]
```diff
@@ -18,8 +20,16 @@
 
   private MultiMap<ClassOrInterfaceType, Sequence> literalMap;
 
-  ClassLiteralExtractor(MultiMap<ClassOrInterfaceType, Sequence> literalMap) {
+  /**
+   * The term frequency mapping is used for the weighted constants option. This will keep track of
+   * how often a constant appears across all constants.
```
Comment (mernst) [inline src/main/java/randoop/reflection/ClassLiteralExtractor.java]: "across all constants": should this be "across all classes", or some other term?  If the current wording is correct, please clarify.

File: src/main/java/randoop/reflection/WeightedConstantsOperationModel.java [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+import randoop.main.ClassNameErrorHandler;
+import randoop.operation.*;
+import randoop.sequence.Sequence;
+import randoop.test.ContractSet;
+import randoop.types.ClassOrInterfaceType;
+import randoop.types.Type;
+import randoop.util.MultiMap;
+
+/**
+ * {@code WeightedConstantsOperationModel} represents the information context from which tests are
```
Comment (mernst) [inline src/main/java/randoop/reflection/WeightedConstantsOperationModel.java]: What is an "information context" and how does it generate tests?

File: src/main/java/randoop/reflection/WeightedConstantsOperationModel.java [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+
+  /** Create an empty model of test context. */
+  private WeightedConstantsOperationModel() {
+    classTypes = new LinkedHashSet<>();
+    inputTypes = new LinkedHashSet<>();
+    classLiteralMap = new MultiMap<>();
+
+    annotatedTestValues = new LinkedHashSet<>();
+    contracts = new ContractSet();
+    contracts.add(EqualsReflexive.getInstance());
```
Comment (mernst) [inline src/main/java/randoop/reflection/WeightedConstantsOperationModel.java]: This code is duplicated exactly in `OperationModel.java` and here is `WeightedConstantsOperationModel.java`.  That will be very error-prone during maintenance and is confusing even now.

It's not just this constructor, though:  much of this class is copy-and-pasted, which might be OK during a quick experiment but is not acceptable for code that would be merged into master.

Can you please eliminate the code duplication?  That will make the code easier to understand, less error-prone during maintenance, and shorter.

File: src/main/java/randoop/util/Randomness.java [hunk 13]
```diff
@@ -90,6 +92,51 @@ public static int nextRandomInt(int i) {
     throw new BugInRandoopException();
   }
 
+  public static <T extends WeightedElement> T randomMemberWeighted(
```
Comment (mernst) [inline src/main/java/randoop/util/Randomness.java]: Every class, method, and field must have documentation giving its specification.  Please add it for the routines in this file.  I can't review code unless I know what it is supposed to do.
(I realize that a few parts of Randoop are missing the documentation that they need, but let's not contribute to that problem.)

File: src/main/java/randoop/util/WeightedList.java [hunk 14]
```diff
@@ -24,23 +30,30 @@ public WeightedElement getRandomElement() {
...
-    double currentPoint = 0;
-    for (int i = 0; i < theList.size(); i++) {
-      currentPoint += theList.get(i).getWeight();
-      if (currentPoint >= randomPoint) {
-        return i;
+    double randomPoint = Randomness.random.nextDouble() * totalWeight;
+    return binarySearchForIndex(randomPoint);
+  }
+
+  private int binarySearchForIndex(double point) {
```
Comment (mernst) [inline src/main/java/randoop/util/WeightedList.java]: Missing documentation here, too.

File: src/main/java/randoop/util/WeightedList.java [hunk 15]
```diff
@@ -2,19 +2,25 @@
 
 import java.util.ArrayList;
 import java.util.List;
-import randoop.BugInRandoopException;
 
 public class WeightedList {
 
   private List<WeightedElement> theList;
+  private List<Double> cumulativeWeights;
```
Comment (mernst) [inline src/main/java/randoop/util/WeightedList.java]: Please document the new fields you added.  It would be a good deed to add documentation elsewhere, especially if that would help in understanding your changes.

File: src/main/java/randoop/util/Randomness.java [hunk 16]
```diff
@@ -90,6 +92,51 @@ public static int nextRandomInt(int i) {
...
+    totalCallsToRandom++;
+    if (Log.isLoggingOn()) {
+      Log.logLine("randoop.util.Randomness: " + totalCallsToRandom + " calls so far.");
+    }
+    double randomPoint = Randomness.random.nextDouble() * max;
+
+    return list.get(binarySearchForIndex(list, cumulativeWeights, randomPoint));
+  }
+
+  public static int binarySearchForIndex(
```
Comment (mernst) [inline src/main/java/randoop/util/Randomness.java]: What is the relationship between the two routines named `binarySearchForIndex` that you added?  Their source code is identical, which is a code smell.  I suggest that you abstract them into a single routine, to avoid code duplication.

**Comment by jkotalik** [comment]
Thanks for the feedback, we realized we didn't fix a few small things, so we will be fixing this ASAP and recreating a new merge request.

**Comment by mernst** [comment]
OK.  Please try to fix the things I noted in my review; that will make the next round go more smoothly!
Thanks for your work on the pull request.

File: src/main/java/randoop/reflection/WeightedConstantsOperationModel.java (hunk 12 above)
Comment (moychal) [inline src/main/java/randoop/reflection/WeightedConstantsOperationModel.java]: We essentially did these copy-paste duplications + created `AbstractOperationModel.java` because we couldn't just extend 'OperationModel.java' without changing its factory constructor.

File: src/main/java/randoop/main/GenTests.java (hunk 8 above)
Comment (moychal) [inline src/main/java/randoop/main/GenTests.java]: Updated in newest commit, thanks for the feedback!

File: src/main/java/randoop/main/GenTests.java (hunk 7 above)
Comment (moychal) [inline src/main/java/randoop/main/GenTests.java]: Updated in newest commit, thanks for the feedback!

File: src/main/java/randoop/generation/ForwardGenerator.java (hunk 2 above)
Comment (moychal) [inline src/main/java/randoop/generation/ForwardGenerator.java]: Updated in newest commit, thanks for the feedback!

File: src/docs/manual/index.html (hunk 4 above)
Comment (moychal) [inline src/docs/manual/index.html]: Should be fixed now, thanks!

File: src/main/java/randoop/main/GenInputsAbstract.java (hunk 6 above)
Comment (moychal) [inline src/main/java/randoop/main/GenInputsAbstract.java]: We have removed the mentions of the name "DigDog".

Would it be more appropriate to explain our command-line additions in depth in the developer manual or here in the descriptions of the command-line options?

File: src/main/java/randoop/reflection/AbstractOperationModel.java (hunk 9 above)
Comment (moychal) [inline src/main/java/randoop/reflection/AbstractOperationModel.java]: Documentation added.

Ambiguities about this class are also in the other comment about code duplication in `WeightedConstantsOperationModel.java`.

File: src/docs/manual/index.html (hunk 4 above)
Comment (moychal) [inline src/docs/manual/index.html]: Question, should we completely remove all the changes?  Including the descriptions generated for our new command-line options?

File: src/main/java/randoop/generation/WeightedComponentManager.java (hunk 3 above)
Comment (moychal) [inline src/main/java/randoop/generation/WeightedComponentManager.java]: Addressed, thanks!

File: src/main/java/randoop/util/WeightedList.java (hunk 15 above)
Comment (moychal) [inline src/main/java/randoop/util/WeightedList.java]: Changes removed (used only in early versions of our project's implementation).  Thanks!

File: src/main/java/randoop/util/Randomness.java (hunk 16 above)
Comment (moychal) [inline src/main/java/randoop/util/Randomness.java]: Fixed, only exists in `Randomness.java` now.

File: src/docs/manual/index.html (hunk 4 above)
Comment (mernst) [inline src/docs/manual/index.html]: Please remove the gratuitous changes, to minimize the size of the diffs.
The descriptions generated for your command-line options are essential to retain!
